src/workerd/api/sql.c++

```diff
@@ -19,6 +19,11 @@ jsg::Ref<SqlStorage::Cursor> SqlStorage::exec(jsg::Lock& js, kj::String querySql
   return jsg::alloc<Cursor>(*sqlite, regulator, querySql, kj::mv(bindings));
 }
 
+kj::StringPtr SqlStorage::ingest(jsg::Lock& js, kj::String querySql) {
+  SqliteDatabase::Regulator& regulator = *this;
+  return sqlite->ingestSql(regulator, querySql);
```

Comment:
Perhaps in the case that nothing is consumed, we should return `null`. Otherwise the string is being copied back and forth from JS (converted UTF-16 -> UTF-8 -> UTF-16) for no reason. Could be expensive if a long SQL statement arrives in many small chunks.

Reply:
If `sqlite->ingestSql` returned null could `SqlStorage::ingest` return the input string unchanged (i.e. without converting back)? So in JS land you can still write:

```js
bufferAfter = sql.ingest(bufferBefore)
```

And you'll just find `bufferAfter === bufferBefore` if no complete statements were present?

Reply:
Alternatively, would it make sense for `sql.ingest` to actually operate on an `ArrayBuffer`, in which case it could go back to returning the _number_ of elements in the input that can now be dropped, but we side-step JS's string encoding?

Is an Response.body's ReadableStream encoded as UTF-16 or UTF-8?

Reply:
This feels like it'd complicate the API quite a bit for a less than common use-case (long SQL statement, small chunks). `SQLITE_LIMIT_SQL_LENGTH` is set to 100k anyway so at worst it'd be a chunk size of, what, 4kb? So it'd go back-and-forth across the JS/C++ boundary 25 times in the absolute worst case?

Reply:
Might I suggest a different approach entirely? Define the API as a `WritableStream`, this way you could pipe into it, then await the results separately...

```js
const ingester = sql.ingester();
ctx.waitUntil(response.body.pipeTo(ingester));
for await (const results of ingester) {
  // handle results of each distinct sql operation piped in...
}
```

This way there would be no need for the additional book keeping boilerplate or the additional copying.

Reply:
> If sqlite->ingestSql returned null could SqlStorage::ingest return the input string unchanged (i.e. without converting back)? So in JS land you can still write:

Theoretically I prefer that as an API, but every time you cross over between JS and C++, the string is being converted between UTF-16 and UTF-8, respectively. So simply returning the same string is not O(1).

Then again, even if `ingest()` returns null when there's nothing to consume, you're still doing the JS -> C++ conversion with every call, only avoiding the conversion back from C++ -> JS. So maybe it's not worth it.

Using ArrayBuffers everywhere could avoid this issue, they are zero-copy. In fact, maybe that ends up being more convenient anyway, given you are receiving the bytes from a stream on the JS side... you can skip the whole TextDecoder step and make things more efficient...

Reply:
Yeah I'm happy with this slight inefficiency for now. I do think I'll add a separate API for consuming a ReadableStream. I really like `for await (const results of ingester)` because we could start reporting back progress as the import is running.

I'm not sure how that would work with the output gate so I think I'd like to punt on making this more optimised/explanatory until I have a full end-to-end example working from Wrangler <-> R2 <-> D1.